[Core] Adding `FromConnectivityGenerateEntities`

loumalouomega opened the pull request:
**📝 Description**

Adding `FromConnectivityGenerateEntities` methods in order to generate conditions/elements from connectivities provided with `std::vector<std::vector<std::size_t>>`. The method works in parallel.

**🆕 Changelog**

- [Adding `FromConnectivityGenerateEntities`](https://github.com/KratosMultiphysics/Kratos/commit/839cdfe6bc3186c67e58e4e07fad98e59d7e0b11)
- [Expose to python](https://github.com/KratosMultiphysics/Kratos/commit/c216b26c8cde70772722846e151b93d1584fc05d)
- [Adding tests](https://github.com/KratosMultiphysics/Kratos/commit/1ee425a00c46df3a4b532cc8d26be001c1ffb5fa)

loumalouomega committed:
[Core] Adding `FromConnectivityGenerateEntities`

loumalouomega committed:
Leftover

loumalouomega committed:
Expose to python

loumalouomega committed:
Adding tests

loumalouomega committed:
Missing

loumalouomega committed:
Missing KRATOS_API(KRATOS_CORE)

sunethwarna commented on kratos/python/add_other_utilities_to_python.cpp:
```diff
@@ -828,6 +829,25 @@ void AddOtherUtilitiesToPython(pybind11::module &m)
+            const Properties::Pointer pProperties) {
+                ModelPartUtils::FromConnectivityGenerateEntities<Condition>(rEntityName, rEntitiesConnectivities, rThisNodes, rThisConditions, pProperties);
+            })
+    ;
```
I would add `py::arg` to the arguments.

loumalouomega replied:
Meaning?

sunethwarna replied:
```suggestion
        .def("FromConnectivityGenerateElements", [](
            const std::string& rEntityName,
            const std::vector<std::vector<std::size_t>>& rEntitiesConnectivities,
            ModelPart::NodesContainerType& rThisNodes,
            ModelPart::ElementsContainerType& rThisElements,
            const Properties::Pointer pProperties) {
                ModelPartUtils::FromConnectivityGenerateEntities<Element>(rEntityName, rEntitiesConnectivities, rThisNodes, rThisElements, pProperties);
            })
        .def("FromConnectivityGenerateConditions", [](
            const std::string& rEntityName,
            const std::vector<std::vector<std::size_t>>& rEntitiesConnectivities,
            ModelPart::NodesContainerType& rThisNodes,
            ModelPart::ConditionsContainerType& rThisConditions,
            const Properties::Pointer pProperties) {
                ModelPartUtils::FromConnectivityGenerateEntities<Condition>(rEntityName, rEntitiesConnectivities, rThisNodes, rThisConditions, pProperties);
            }, py::arg("condition_name"), py::arg("condition_connectivities_list"), ...)
    ;
```

loumalouomega replied:
Ah, ok

loumalouomega committed:
Correct test

loumalouomega committed:
Add `py::arg`

loumalouomega committed:
Rename to `FromConnectivityGenerateEntities` by suggestion of @rubenzorrilla

loumalouomega committed:
Merge branch 'master' into core/connectivity-element-generator-utility

loumalouomega committed:
Revert "Correct test"

This reverts commit 4a6a652a31a08568812479f721091c957868e0a5.

loumalouomega committed:
Sort in order to fix test

roigcarlo committed:
Trying to fix ci.

roigcarlo committed:
Trying to respect intel flags

roigcarlo committed:
Hopefully fixing CI

What a beautiful world we live on in which we can experience the joy of compiling for multiple platforms and operating systems

roigcarlo committed:
Missing if and commenting candelier test

loumalouomega committed:
Minor clean up

loumalouomega committed:
In order to satisfy MSVC I removed the explicit instantation because linking problems

loumalouomega committed:
Merge branch 'master' into core/connectivity-element-generator-utility

philbucher commented on kratos/python/add_other_utilities_to_python.cpp:
```diff
@@ -828,6 +829,36 @@ void AddOtherUtilitiesToPython(pybind11::module &m)
         .def("CreateSphereTriangularMesh", &CppTestsUtilities::CreateSphereTriangularMesh)
     ;
 
+    m.def_submodule("ModelPartUtils", "Auxiliary utilities for model parts.")
```
Its a class, not a namespace, this is wrong

loumalouomega replied:
Okay

philbucher reviewed (commented):
Conceptually, this should be used by the ConnPreserveModeller no?

loumalouomega committed:
Define as class

pooyan-dadvand commented on kratos/utilities/model_part_utils.h:
```diff
@@ -92,9 +97,144 @@ class ModelPartUtils
+        const std::size_t number_of_entities = rEntitiesConnectivities.size();
+        rThisEntities.reserve(current_size + number_of_entities);
+
+        // We assume that the first entity in the list is the one we want to clone
```
Is the comment correct?

loumalouomega replied:
Yes, `reserve` works with the destiny capacity.

pooyan-dadvand replied:
But you are not using the first entity as the one you want to clone. You are getting it from the input.

loumalouomega replied:
The vector may be empty, this way you ensure you are using the element you are interested. If you want I can add a method with a prototype

loumalouomega replied:
Ah, you mean the comment, it is a leftover froma  previous iteration, cleaning

loumalouomega replied:
Done

pooyan-dadvand reviewed (commented):
Just a minor comment

loumalouomega committed:
Merge branch 'master' into core/connectivity-element-generator-utility

loumalouomega committed:
Adding python test

loumalouomega committed:
Merge branch 'master' into core/connectivity-element-generator-utility

loumalouomega committed:
Remove leftover comments and add direct method for prototypes

loumalouomega committed:
Merge branch 'master' into core/connectivity-element-generator-utility
